modules/join/src/main/java/org/apache/ignite/internal/join/ClusterManagementGroupManager.java

```diff
@@ -0,0 +1,198 @@
+    public ClusterManagementGroupManager(ClusterService clusterService, Loza raftManager, RestModule restModule) {
+        this.clusterService = clusterService;
+        this.raftManager = raftManager;
+        this.restModule = restModule;
+        this.clusterInitializer = new ClusterInitializer(clusterService);
+
+        MessagingService messagingService = clusterService.messagingService();
+
+        var msgFactory = new InitMessagesFactory();
+
+        messagingService.addMessageHandler(InitMessageGroup.class, (msg, addr, correlationId) -> {
+            if (!busyLock.enterBusy()) {
+                if (correlationId != null) {
+                    messagingService.respond(addr, errorResponse(msgFactory, new NodeStoppingException()), correlationId);
+                }
+
+                return;
+            }
+
+            try {
+                if (msg instanceof CancelInitMessage) {
+                    log.info("CMG initialization cancelled, reason: " + ((CancelInitMessage) msg).reason());
+
+                    raftManager.stopRaftGroup(CMG_RAFT_GROUP_NAME);
+
+                    // TODO: drop the Raft storage as well, https://issues.apache.org/jira/browse/IGNITE-16471
+                } else if (msg instanceof CmgInitMessage) {
+                    assert correlationId != null;
+
+                    String[] nodeIds = ((CmgInitMessage) msg).cmgNodes();
+
+                    List<ClusterNode> nodes = resolveNodes(clusterService, Arrays.asList(nodeIds));
+
+                    raftManager.prepareRaftGroup(CMG_RAFT_GROUP_NAME, nodes, CmgRaftGroupListener::new)
+                            .whenComplete((service, e) -> {
+                                if (e == null) {
+                                    Peer leader = service.leader();
+
+                                    assert leader != null;
+
+                                    messagingService.respond(addr, leaderElectedResponse(msgFactory, leader), correlationId);
+                                } else {
+                                    messagingService.respond(addr, errorResponse(msgFactory, e), correlationId);
+                                }
+                            });
+                }
+            } catch (Exception e) {
+                log.error("Exception when initializing the CMG", e);
+
+                if (correlationId != null) {
+                    messagingService.respond(addr, errorResponse(msgFactory, e), correlationId);
+                }
+            } finally {
+                busyLock.leaveBusy();
+            }
+        });
+    }
+
+    /**
+     * Initializes the cluster that this node is present in.
+     *
+     * @param metaStorageNodeNames names of nodes that will host the Meta Storage.
+     * @param cmgNodeNames names of nodes that will host the Cluster Management Group.
+     */
+    public Leaders initCluster(Collection<String> metaStorageNodeNames, Collection<String> cmgNodeNames) throws NodeStoppingException {
```

Comment:
It seems that this method does not belong to this class. The class is about CMG, but the method is about the whole initialization (which also includes meta-storage initialization).

I suggest to create another class which responsibility would be to perform the initialization as a whole; this method would be moved to that class (maybe with the initialization of a REST module, or maybe not, because a REST handler and a service seem to be in different layers).

Reply:
Same here, cluster initialization can be moved to a separate component but is managing yet another component worth it? Especially since we don't have any dependency injection mechanism...